feat(antnode): simulate replication and check if records are being st…

RolandSherwin opened the pull request:
- A simulation to test if our replication is working as intended and check if the records are being stored at the right location after uploading.

RolandSherwin committed:
feat(antnode): simulate replication and check if records are being stored at the right location

maqi commented on ant-node/tests/simulate_replication.rs:
```diff
@@ -0,0 +1,1183 @@
+    for i in 0..num_chunks {
+        // Generate random chunk
+        let chunk_addr = NetworkAddress::ChunkAddress(ChunkAddress::new(XorName::random(&mut rng)));
+        let chunk_size = 1024 * 1024; // 1MB
```
to make the simulation quicker, the `content` can be very trival.
what matters is the `address` only

RolandSherwin replied:
The chunk size is only used inside quote generation. Maybe I'll rearrange things.

maqi commented on ant-node/tests/simulate_replication.rs:
```diff
@@ -0,0 +1,1220 @@
+        }
+
+        // Check 2: Get K_VALUE (20) closest peers from our routing table to the data address
+        let closest_k = self.find_closest_local(record_addr, LIBP2P_K_VALUE);
```
this `find_closest_local` returns K peers closest to the target but exculding self
which mens the later on `out_of_k_payees` will still containing self.

this will always trigger the later on `responsible_distance` check, which being more strict and different to the production behaviour

maqi reviewed (changes requested):
Put in `Request changes` as showing different result to the testnet runs

maqi commented on ant-node/tests/simulate_replication.rs:
```diff
@@ -0,0 +1,1373 @@
+
+        // Select payment targets based on mode
+        let paid_nodes: Vec<PeerId> = match payment_mode {
+            PaymentMode::SingleNode => {
```
all candidates that got picked in the payment, even they got 0 paid, shall still accept the upload.
hence here, all 5 shall be included in the payment. otherwise will result in rejection of `NotInPaymentList`

maqi reviewed (changes requested):
best make the test to be executed as a separate CI run, so that the results can then be seen more explicitly.

RolandSherwin committed:
feat(ci): run simulate_replication test during CI

RolandSherwin committed:
refactor(simulation): sync replication logic with production code

- Replace middle-range fallback in is_in_range with simple farthest peer
  distance comparison to match production behavior
- Add majority accumulation requiring 2+ peers to report same record
  before storing (matches production consensus requirement)
- Fix closest peers self-inclusion to exclude self where production does
- Remove unused REPLICATION_FACTOR constant
- Update statistics output with simplified in-range metrics and new
  majority accumulation tracking

RolandSherwin committed:
feat(simulation): add JSON export and ASCII visualization

- Add SimulationExport structs for JSON serialization of results
- Export simulation config, round data, and coverage stats to JSON
- Add ASCII progress bars for coverage visualization
- Add coverage distribution histogram with bar charts
- Add storage distribution percentile visualization (min/p25/median/p75/max)

RolandSherwin committed:
feat(simulation): add node failure injection and fault tolerance metrics

- Add configurable failure rate (10% default) to simulate node failures
- Inject failures between replication and verification phases
- Calculate fault tolerance metrics: data loss rate, critical risk rate,
  mean coverage, std dev, p1 coverage, and theoretical survival probability
- Include fault tolerance metrics in JSON export and terminal output

RolandSherwin committed:
refactor(simulation): split test into modular structure

Split the large simulate_replication.rs file into organized modules:
- config.rs: Constants and simulation configuration
- counters.rs: Global atomic counters for statistics
- types.rs: Core data structures and ReplicationAccumulator
- node.rs: SimulatedNode implementation
- stats.rs: UploadStats and CoverageStats helpers
- export.rs: JSON export structures
- mod.rs: Main test function and report generation

RolandSherwin committed:
feat(simulation): add Monte Carlo mode for statistical confidence

Add multi-trial simulation capability with 95% confidence intervals:
- Extract single trial into reusable run_single_trial() function
- Add run_monte_carlo_simulation() that runs configurable trials
- Export results to monte_carlo_results.json with CI statistics
- Reset counters between trials for accurate per-trial metrics
- Restructure test modules for proper Rust 2018+ module paths

maqi commented:
Simulation somehow outdated and no longer necessary.
